Add minimal support for EK-RA8M1 (CM85)

thaoluonguw opened the pull request:
* Add support for Arm Cortex-M85
* Add minimal support for EK-RA8M1 as following

| Interface | Controller | Driver/Component     |
| ----------- | ----------- | ----------- |
| GPIO      | on-chip    | gpio  |
| MPU       | on-chip    | arch/arm |
| NVIC      | on-chip    | arch/arm  |
| UART      | on-chip    | serial |

Need: https://github.com/zephyrproject-rtos/hal_renesas/pull/17

duynguyenxa committed:
arch: arm: Add initial support for Cortex-M85 Core

Add initial support for the Cortex-M85 Core which is an implementation
of the Armv8.1-M mainline architecture.

The support is based on the Cortex-M55 support that already exists in
Zephyr.

Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>

duynguyenxa committed:
soc: renesas: ra: Add initial support for RA8M1 SOC series

Add minimal support for RA8M1 SOC series.

Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>
Signed-off-by: Quy Tran <quy.tran.pz@renesas.com>

duynguyenxa committed:
drivers: pinctrl: Add pinctrl driver for RA8 series

This is the initial commit to support minimum pinctrl
driver for Renesas MCU RA8M1.

Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>

duynguyenxa committed:
driver: gpio: Add initial gpio drirver support for RA8M1

This is the initial commit to support for gpio driver
for RA8M1 MCU, the coding is base on renesas fsp hal

Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>

duynguyenxa committed:
boards: arm: Add initial support for EKRA8M1

Add support for the EK RA8M1 board
This board is using Renesas RA8M1 MCU.

Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>
Signed-off-by: The Nguyen <the.nguyen.yf@renesas.com>

thenguyenyf committed:
driver: serial: Add serial driver support for Renesas RA8 devices

Add implementation of sci_b_uart for Renesas RA device

Signed-off-by: The Thanh. Nguyen <the.nguyen.yf@renesas.com>

soburi commented:
@thaoluonguw 

Thank you for your contribution.

First, you seem to be affiliated with Renesas company, so let me confirm that first.

Q. The FSP license change to Apache2?

I think it was Renesas' own license when I saw it before, but has that changed recently?

carlescufi commented on west.yml:
```diff
@@ -23,6 +23,8 @@ manifest:
       url-base: https://github.com/zephyrproject-rtos
     - name: babblesim
       url-base: https://github.com/BabbleSim
+    - name: renesas
```
Please don't do this. Instead, open a Pull Request in the Zephyr hal_renesas repo and use `pull/<PR number>/head` as the revision.

thaoluonguw replied:
@carlescufi : Thank you so much for your feedback. We fixed it. Please check.

soburi commented on include/zephyr/dt-bindings/pinctrl/renesas/ra-pinctrl.h:
```diff
@@ -0,0 +1,56 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef __RA_PINCTRL_H__
```
A full-path-based name is preferred.

thaoluonguw replied:
@soburi : Thank you for your feedback. We updated it. Please check.

soburi commented on drivers/serial/uart_renesas_ra8_sci_b.c:
```diff
@@ -0,0 +1,1164 @@
+	return 0;
+}
+
+static void uart_ra_sci_b_poll_out(const struct device *dev, unsigned char c)
```
One confirmation.

Is the `samples/drivers/uart/echo_bot/` sample work correctly?
This sample uses polling for sending and uses isr for receiving.

thaoluonguw replied:
@soburi : Yes, we confirmed samples/drivers/uart/echo_bot/ sample work correctly.

soburi replied:
I got it. (I got stuck when I tried it with RA4M1, so I checked it.)

soburi reviewed (changes requested):
[Merge branch 'zephyrproject-rtos:main' into ra8-support-1.2](https://github.com/zephyrproject-rtos/zephyr/pull/72150/commits/7e781803596f84706cf2721f6c8f70431f78cf26)

Don't merge. Please use rebase.

soburi commented on soc/renesas/ra/ra8/bsp_mcu_family_cfg.h:
```diff
@@ -0,0 +1,571 @@
+#define BSP_MCU_FAMILY_CFG_H_
+#include "bsp_mcu_device_pn_cfg.h"
+#include "bsp_mcu_device_cfg.h"
+#include "../../../ra/fsp/src/bsp/mcu/ra8m1/bsp_override.h"
```
Is it okay to configure the folder as `soc/renesas/ra/ra8`?
Looking at this line, I think `soc/renesas/ra/ra8m1` is better.

thaoluonguw replied:
@soburi : Thank you for your comment. We updated it. Currently, only ra8m1, we just keep "#include "bsp_override.h"
When others RA8 MCU are available, it will be:
```
#if defined(CONFIG_SOC_R7FA8M1AHECBD)
#include "ra8m1/bsp_override.h"
#include "ra8m1/bsp_mcu_info.h"
#define BSP_MCU_GROUP_RA8M1 (1)
#elif defined(CONFIG_SOC_R7FA8D1BHECBD)
#include "ra8d1/bsp_override.h"
#include "ra8d1/bsp_mcu_info.h"
```
Corresponding in hal_renesas, "fsp/ra/fsp/src/bsp/mcu" will be added into zephyr_include_directories

soburi commented on soc/renesas/ra/ra8/Kconfig.soc:
```diff
@@ -0,0 +1,22 @@
+# SPDX-License-Identifier: Apache-2.0
+
+
+config SOC_SERIES_RA8
+	bool
+	select SOC_FAMILY_RENESAS_RA
+	help
+		Renesas RA8 series
+
```
Even if the Renesas RA series belongs to the same series in the catalog, the configuration may vary greatly. 
For example, the peripheral register configuration of the RA4M1 is very different from that of the RA4M2 and is rather similar to the RA6M1.
(I have no idea why it's like this...)

For this reason, it seems better to divide the series here into groups on the catalog (RA8M1, RA6E1, ...) rather than series on the catalog (RA8, RA6, ...).

thaoluonguw replied:
@soburi : 

Our first intention when group series on the catalog (RA8, RA6, ...) is to avoid some duplicated codes (as you can see, in RA8 series, there are a few differences between RA8M1, RA8D1 and RA8T1) and follow Renesas RA Family MCU Portfolio (described at: https://www.renesas.com/us/en/products/microcontrollers-microprocessors/ra-cortex-m-mcus).

We also agree with you that it will be difficult to deduce unify rules for this "group". It will make codes become complicated. So, after discussion, we made decision to update as your proposal. Please check the update and thank you so much.

soburi commented on dts/bindings/gpio/renesas,ra8-gpio.yaml:
```diff
@@ -0,0 +1,23 @@
+  reg:
+    required: true
+
+  port:
```
This is able to determine by address offset

thaoluonguw replied:
@soburi : Thank you for your advice. We saw your method for this case.
First, may I confirm that there is any special improvement when change to use this method?
Because we are supporting for new board, we will get information from HWM to update for dts (include port information is OK). When change to calculation offset method, can it impact to the latency?

thaoluonguw replied:
@soburi : Thank you very much for your comment. We understood. Because it's not blocking, let's us keep as current implementation for easy checking with HWM.

thaoluonguw commented:
> @thaoluonguw
> 
> Thank you for your contribution. First, you seem to be affiliated with Renesas company, so let me confirm that first.
> 
> Q. The FSP license change to Apache2?
> 
> I think it was Renesas' own license when I saw it before, but has that changed recently?
> 
> Also, I think the RA series interrupt (ICU) operates by dynamically allocating an IRQ. In my understanding, the FSP automatically generates the code for interrupt allocation, but please let me know if you have any thoughts on how to deal with this issue. I think it will be easier to review once I understand it.

Hello @soburi , thank you very much for your feedback and I am so sorry for my late responding.

> Q. The FSP license change to Apache2?
> I think it was Renesas' own license when I saw it before, but has that changed recently?

Yes, FSP license changed to BSD-3-Clause from FSPv5.3.0

>the FSP automatically generates the code for interrupt allocation, but please let me know if you have any thoughts on how to deal with this issue

Currently, we use fixed interrupts which defined in dts. We are considering for better method.

thaoluonguw commented on dts/arm/renesas/ra/ra8/ra8x1.dtsi:
```diff
@@ -0,0 +1,241 @@
+			status = "disabled";
+		};
+
+		sci9: sci9@40358900 {
+			compatible = "renesas,ra8-uart-sci-b";
+			channel = <9>;
+			interrupts = <24 1>, <25 1>, <26 1>, <27 1>;
+			interrupt-names = "rxi", "txi", "tei", "eri";
+			reg = <0x40358900 0x100>;
+			status = "disabled";
+		};
```
@soburi : Thank you for your feedback. Currently, we haven't planned to support I2C, SPI and Smart card interface on the SCI.

soburi replied:
Since devicetree is a hardware description, it should be written based on hardware configuration, not support status.
Since SCI is a common device with several communication methods to choose from, such a devicetree representation should be given.

NXP's FlexComm has functionality very similar to SCI, but even in that case, functions are split into child nodes, so SCI is also necessary to separate the common parts of shared part and each function.

https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/arm/nxp/nxp_mcxn94x_common.dtsi#L173-L206

thaoluonguw commented:
> [Merge branch 'zephyrproject-rtos:main' into ra8-support-1.2](https://github.com/zephyrproject-rtos/zephyr/pull/72150/commits/7e781803596f84706cf2721f6c8f70431f78cf26)
> 
> Don't merge. Please use rebase.

@soburi : Thank you for your advice. We fixed for it. Please check.

ydamigos commented on drivers/gpio/gpio_renesas_ra8.c:
```diff
@@ -0,0 +1,244 @@
+		R_SYSTEM->PRCR = ((R_SYSTEM->PRCR | PRCR_KEY) & (uint16_t)~R_SYSTEM_PRCR_PRC1_Msk);
+	}
+
+#if BSP_CFG_MCU_PART_SERIES == 8
```
Why are we not using a Zephyr definition like SOC_SERIES_RA8?

thaoluonguw replied:
@ydamigos : Thank you for your feedback. We will change to use Zephyr macro. However, we found that this is redundant codes in this file which support for RA8 only. We removed these redundant codes. Please check. Thank you.

ydamigos commented on boards/renesas/ek_ra8m1/doc/index.rst:
```diff
@@ -0,0 +1,171 @@
+:ref:`build_an_application` and :ref:`application_run` for more details on
+building and running.
+
+**Note:** Currently Zephyr SDK v0.16.5 hasn't supported for Cortex Arm-M85 build. To build for EK-RA8M1
```
Note needs an update to mention that Zephyr SDK v0.16.6 is needed.

thaoluonguw replied:
@ydamigos : Thank you so much for your feedback. We updated it. Please check.

ydamigos commented on soc/renesas/ra/ra8/bsp_mcu_device_pn_cfg.h:
All these "*_cfg.h" files are configuring FSP, aren't they? I propose to move them in hal_renesas repository and use DT values to define the macros (where it is possible).

thaoluonguw replied:
@ydamigos : We moved them to hal_renesas repository. Thank you so much for your proposal.

soburi replied:
@ydamigos 

"*_cfg.h" contains many refer to the devicetree.
I think it makes reverse references.
So, these files are better to place in the zephyr tree.
I want to hear your opinion.

ydamigos replied:
@soburi Those files configure FSP. For me, they should be placed in hal_renesas and their macros should be populated using DT macros where is needed. Other vendors are doing the same, don't they?

soburi replied:
@ydamigos 
Thanks for your comment. You're right. Let's keep going.

soburi commented on soc/renesas/ra/ra8m1/soc.c:
```diff
@@ -0,0 +1,56 @@
+ *
+ * @return 0
+ */
+static int renesas_ra8_init(void)
```
Symbols and comments still remain "RA8".
Please change to RA8M1.

soburi commented on soc/renesas/ra/ra8m1/Kconfig.soc:
```diff
@@ -0,0 +1,22 @@
+	bool
+	select SOC_FAMILY_RENESAS_RA
+	help
+		Renesas RA8M1 series
```
help description usually uses a 2-space indentation.

soburi commented on dts/arm/renesas/ra/ra8/r7fa8m1ahecbd.dtsi:
```diff
@@ -0,0 +1,100 @@
+#include <zephyr/dt-bindings/clock/ra_clock.h>
+
+/ {
+	clocks {
```
As I understand it, the only difference in the branch numbers is the flash size.
Wouldn't it be better to include the clock in `dts/arm/renesas/ra/ra8/ra8x1.dtsi`?

thaoluonguw replied:
dts/arm/renesas/ra/ra8/ra8x1.dtsi is target for RA8M1, RA8D1 and RA8T1. Clock will be different between them. We will consider moving to r7fa8m1xh.dtsi.

thaoluonguw replied:
We updated it. Please check. Thank you.

soburi commented on west.yml:
```diff
@@ -210,7 +210,7 @@ manifest:
     - name: hal_renesas
       path: modules/hal/renesas
-      revision: b4fe8925d95bddc27e6d1666890f560d30cf9166
+      revision: pull/17/head
```
Please add [DNM] at the beginning of your commit message.
[manifest: Correct commit ID for hal_renesas](https://github.com/zephyrproject-rtos/zephyr/pull/72150/commits/be13c606e6374824d274548e6524d50f93c315c1)

You will rewrite this commit after hal_renesas is updated.

soburi commented on dts/arm/renesas/ra/ra8/r7fa8m1ahecbd.dtsi:
```diff
@@ -0,0 +1,100 @@
+			fclk-div = <RA_SYS_CLOCK_DIV_8>;
+			sdclk-out = <1>;
+
+			uclk-source = <RA_CLOCK_SOURCE_DISABLE>;
+			uclk-div = <RA_USB_CLOCK_DIV_5>;
+			u60clk-source = <RA_CLOCK_SOURCE_DISABLE>;
+			u60clk-div = <RA_USB60_CLOCK_DIV_5>;
+			octaspiclk-source = <RA_CLOCK_SOURCE_DISABLE>;
+			octaspiclk-div = <RA_OCTA_CLOCK_DIV_4>;
+			canfdclk-source = <RA_CLOCK_SOURCE_DISABLE>;
+			canfdclk-div = <RA_CANFD_CLOCK_DIV_8>;
+			clkout-source = <RA_PLL_SOURCE_LOCO>;
+			clkout-div = <RA_SYS_CLOCK_DIV_1>;
+			sciclk-source = <RA_CLOCK_SOURCE_DISABLE>;
+			sciclk-div = <RA_SCI_CLOCK_DIV_4>;
+			spiclk-source = <RA_CLOCK_SOURCE_DISABLE>;
+			spiclk-div = <RA_SPI_CLOCK_DIV_4>;
+			i3cclk-source = <RA_CLOCK_SOURCE_DISABLE>;
+			i3cclk-div = <RA_I3C_CLOCK_DIV_3>;
```
It is easier to understand if you express invalidity by the presence or absence of the property itself rather than set DISABLED.
It is better not to make the property required so that you can omit specifying properties that are related to the disabled clock.

thaoluonguw replied:
Currently, it is necessary when use with FSP (it is referred at [zephyr/ra/ra_cfg/fsp_cfg/bsp/ra8m1/bsp_clock_cfg.h](https://github.com/renesas/hal_renesas/blob/ra8-support-1.2/zephyr/ra/ra_cfg/fsp_cfg/bsp/ra8m1/bsp_clock_cfg.h)).

thaoluonguw replied:
@soburi : We updated it. Please check. Thank you.
FYI>We separated clock into nodes and followed information in "Clock" tab which generated in e2studio.
For unused clocks, it will have status "disabled". We see similar support in other vendors (ex. stm32 - https://github.com/renesas/zephyr/blob/main/dts/arm/st/f0/stm32f0.dtsi)

soburi replied:
Thank you for updating. I think this will allow us to flexibly support other models as well.

soburi commented on dts/arm/renesas/ra/ra8/ra8x1.dtsi:
```diff
@@ -0,0 +1,241 @@
+			status = "okay";
+		};
+
+		pinctrl: pin-contrller@40400800 {
```
controller

soburi commented on soc/renesas/ra/ra8m1/linker.ld:
```diff
@@ -0,0 +1,608 @@
+/*
```
Please rewrite with `zephyr_linker_sources`
https://github.com/zephyrproject-rtos/zephyr/blob/c23132afdf33dc474bc814d4ccd576e76ee6618c/soc/st/stm32/stm32h7x/CMakeLists.txt#L9

and 

`zephyr,memory-region`

https://github.com/zephyrproject-rtos/zephyr/blob/c23132afdf33dc474bc814d4ccd576e76ee6618c/dts/arm/nxp/nxp_rt11xx_cm4.dtsi#L30-L34

You probably don't need to create a new linker.ld.

soburi commented on drivers/gpio/gpio_renesas_ra8.c:
```diff
@@ -0,0 +1,234 @@
+	return 0;
+}
+
+#define GPIO_DEVICE_INIT(__node, __port_num, __suffix, __addr)                                     \
```
Prefix `__` is reserved for a compiler or standard library.
Please don't use here.

soburi commented on drivers/gpio/gpio_renesas_ra8.c:
```diff
@@ -0,0 +1,234 @@
+	4U,
+};
+
+#define RA8_INVALID_PORT_ADDR 0UL
+
+#define RA8_GPIO_PORT_ADDR(nodelabel) \
+	COND_CODE_1(DT_NODE_EXISTS(DT_NODELABEL(nodelabel)), \
+	(DT_REG_ADDR(DT_NODELABEL(nodelabel))), (RA8_INVALID_PORT_ADDR))
+
```
Not referenced.
Please remove.

soburi commented on drivers/gpio/gpio_renesas_ra8.c:
```diff
@@ -0,0 +1,234 @@
+#include <soc.h>
+
+#define PRCR_KEY   0xA500U
+#define VBATT_PORT 4
+
+static const gpio_pin_t vbatt_pins[] = {
+	2U,
+	3U,
+	4U,
+};
```
It seems to be a hardware configuration.
It is good to define in devicetree.

soburi replied:
It is good to define an array of `vbatt_pins` in the devicetree and reference that.

soburi commented on drivers/gpio/gpio_renesas_ra8.c:
```diff
@@ -0,0 +1,234 @@
+#include <zephyr/irq.h>
+#include <soc.h>
+
+#define PRCR_KEY   0xA500U
```
Please clarify what the constants are.

soburi commented on drivers/serial/uart_renesas_ra8_sci_b.c:
```diff
@@ -0,0 +1,1165 @@
+#include "r_transfer_api.h"
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(ra8_uart_sci);
```
The uart_sci will need support in the future.
The name should be `ra8_uart_sci_b` here.

soburi commented on boards/renesas/ek_ra8m1/ek_ra8m1-pinctrl.dtsi:
```diff
@@ -0,0 +1,56 @@
+&pinctrl {
+	status = "okay";
+
+	sci0_default: sci0_default {
+		group1 {
+			/* tx */
+			psels = <RA_PSEL(RA_PSEL_SCI_0, 6, 9)>;
+			drive-strength = "medium";
+		};
+		group2 {
+			/* rx */
+			psels = <RA_PSEL(RA_PSEL_SCI_0, 6, 10)>;
+		};
+	};
+
+	sci2_default: sci2_default {
+		group1 {
+			/* tx */
+			psels = <RA_PSEL(RA_PSEL_SCI_2, 10, 3)>;
+			drive-strength = "medium";
+		};
+		group2 {
+			/* rx */
+			psels = <RA_PSEL(RA_PSEL_SCI_2, 10, 2)>;
+		};
+	};
+
+	sci3_default: sci3_default {
+		group1 {
+			/* tx */
+			psels = <RA_PSEL(RA_PSEL_SCI_3, 3, 10)>;
+			drive-strength = "medium";
+		};
+		group2 {
+			/* rx */
+			psels = <RA_PSEL(RA_PSEL_SCI_3, 3, 9)>;
+		};
+	};
+
```
Please remove pinctrl settings for peripherals that are disabled by default.

thaoluonguw replied:
This is support for all available SCI on EK-RA8M1. We keep it for user easy to refer.

soburi replied:
This is a Zephyr convention, so please support this.
Normally pins can be configured for many different functions, so it is a convention not to force or recommend any particular function as the default.

soburi commented on dts/arm/renesas/ra/ra8/ra8x1.dtsi:
```diff
@@ -0,0 +1,241 @@
+		pinctrl: pin-contrller@40400800 {
+			compatible = "renesas,ra8-pinctrl";
+			reg = <0x40400800 0x3c0>;
+			status = "disabled";
```
I think pinctrl is enabled by default and there is no problem.

soburi commented on drivers/gpio/gpio_renesas_ra8.c:
```diff
@@ -0,0 +1,234 @@
+		R_SYSTEM->PRCR = ((R_SYSTEM->PRCR | PRCR_KEY) & (uint16_t)~R_SYSTEM_PRCR_PRC1_Msk);
+	}
+
+	R_PMISC->PWPRS = 0;
+	R_PMISC->PWPRS = BIT(R_PMISC_PWPR_PFSWE_Pos);
+
+	uint32_t value = R_PFS->PORT[config->port_num].PIN[pin].PmnPFS;
+
+	/* Change mode to general IO mode and disable IRQ and Analog input */
+	WRITE_BIT(value, R_PFS_PORT_PIN_PmnPFS_PMR_Pos, 0);
+	WRITE_BIT(value, R_PFS_PORT_PIN_PmnPFS_ASEL_Pos, 0);
+	WRITE_BIT(value, R_PFS_PORT_PIN_PmnPFS_ISEL_Pos, 0);
+
+	if ((flags & GPIO_OUTPUT) != 0U) {
+		/* Set output pin initial value */
+		if ((flags & GPIO_OUTPUT_INIT_LOW) != 0U) {
+			WRITE_BIT(value, R_PFS_PORT_PIN_PmnPFS_PODR_Pos, 0);
+		} else if ((flags & GPIO_OUTPUT_INIT_HIGH) != 0U) {
+			WRITE_BIT(value, R_PFS_PORT_PIN_PmnPFS_PODR_Pos, 1);
+		}
+
+		WRITE_BIT(value, R_PFS_PORT_PIN_PmnPFS_PDR_Pos, 1);
+	} else {
+		WRITE_BIT(value, R_PFS_PORT_PIN_PmnPFS_PDR_Pos, 0);
+	}
+
+	R_PFS->PORT[config->port_num].PIN[pin].PmnPFS = value;
+
+	R_PMISC->PWPRS = 0;
+	R_PMISC->PWPRS = BIT(R_PMISC_PWPR_B0WI_Pos);
+
```
It is better to use the `pinctrl_configure_pins` function.

soburi commented on drivers/gpio/gpio_renesas_ra8.c:
```diff
@@ -0,0 +1,234 @@
+#endif
+};
+
+static int gpio_ra8_init(const struct device *dev)
```
`DEVICE_DT_DEFINE` accepts NULL as initializer function pointer.
This function can be removed.

soburi commented on drivers/gpio/gpio_renesas_ra8.c:
```diff
@@ -0,0 +1,234 @@
+	R_PMISC->PWPRS = 0;
+	R_PMISC->PWPRS = BIT(R_PMISC_PWPR_PFSWE_Pos);
+
+	uint32_t value = R_PFS->PORT[config->port_num].PIN[pin].PmnPFS;
```
It should be possible to determine, regardless of past settings.

soburi commented on drivers/gpio/gpio_renesas_ra8.c:
```diff
@@ -0,0 +1,234 @@
+	WRITE_BIT(value, R_PFS_PORT_PIN_PmnPFS_ASEL_Pos, 0);
+	WRITE_BIT(value, R_PFS_PORT_PIN_PmnPFS_ISEL_Pos, 0);
+
+	if ((flags & GPIO_OUTPUT) != 0U) {
```
Please handle all properties.
https://github.com/zephyrproject-rtos/zephyr/pull/72150/files#diff-e4af11f77a8bbf03f9175f29acf04a4f2adc7eed434ef6ecfe4470fe66037b88R84-R88

thaoluonguw replied:
We  handle for all properties. Please check. Thank you.

soburi commented on drivers/pinctrl/renesas/ra/pinctrl_renesas_ra8.c:
```diff
@@ -0,0 +1,37 @@
+
+int pinctrl_configure_pins(const pinctrl_soc_pin_t *pins, uint8_t pin_cnt, uintptr_t reg)
+{
+	R_PMISC->PWPRS = 0;
```
Can't I use R_BSP_PinAccessEnable here?

soburi commented on drivers/pinctrl/renesas/ra/pinctrl_renesas_ra8.c:
```diff
@@ -0,0 +1,37 @@
+		 */
+		if ((pin->cfg & BIT(R_PFS_PORT_PIN_PmnPFS_PMR_Pos)) > 0) {
+			/* Clear PMR */
+			R_PFS->PORT[pin->port_num].PIN[pin->pin_num].PmnPFS_b.PMR = 0;
```
There are various ways to do this, but
`R_PFS` is a macro, which makes debugging difficult.
The Zephyr-friendly method is to reference registers defined in the device tree via config, as is done with GPIO.

https://github.com/zephyrproject-rtos/zephyr/pull/72150/files#diff-4fd9e45da58f81a8a9761aefd7ed42633287f99d1ad17a6b230bf0fdfa73681dR176

As pointed out above, using the BSP functions is also an option.

Either way, it's best to make your policy clear.
In my opinion, I recommend the Zephyr method, which uses a device tree.

soburi commented on dts/bindings/serial/renesas,ra8-uart-sci-b.yaml:
```diff
@@ -0,0 +1,34 @@
+    type: int
+    required: true
+
+  tx-dtc:
+    type: boolean
+
+  rx-dtc:
+    type: boolean
```
Please add descriptions.

soburi commented on drivers/gpio/gpio_renesas_ra8.c:
```diff
@@ -0,0 +1,234 @@
+	.port_toggle_bits = gpio_ra8_port_toggle_bits,
+	.pin_interrupt_configure = NULL,
+	.manage_callback = NULL,
+#ifdef CONFIG_GPIO_GET_DIRECTION
+	.port_get_direction = gpio_ra8_port_get_direction,
+#endif
```
`gpio_ra8_port_get_direction` is not available.
Please remove these lines.

soburi commented on drivers/serial/Kconfig.renesas_ra8:
```diff
@@ -0,0 +1,25 @@
+
+if UART_RA8_SCI_B && (UART_INTERRUPT_DRIVEN || UART_ASYNC_API)
+
+config UART_RA_SCI_B_UART_FIFO_ENABLE
```
In the uart_sci (not uart_sci_b) case, There are mixtures of sci with and without FIFO.
It would be appreciated if you could set this for each node in the device tree to keep the config format similar.

soburi commented on drivers/serial/uart_renesas_ra8_sci_b.c:
```diff
@@ -0,0 +1,1165 @@
+};
+
+struct uart_ra_sci_b_data {
+	const struct device *dev;
```
`struct device` is usually placed on config structure.

thaoluonguw replied:
We need it to handle for Asynchronous case. We saw that some others have similar implementation (ex. drivers\serial\uart_stm32 and drivers\serial\uart_ns16550)

soburi commented on dts/arm/renesas/ra/ra8/ra8x1.dtsi:
```diff
@@ -0,0 +1,241 @@
+/*
```
If there is a possibility that RA8M2 or similar will appear in the future, ra8xx.dtsi would be better.

thaoluonguw replied:
We cannot decide it at this time. Let us keep as current and consider it when 2nd generation is available.

soburi commented on drivers/serial/uart_renesas_ra8_sci_b.c:
```diff
@@ -0,0 +1,1165 @@
+static int uart_ra_sci_b_poll_in(const struct device *dev, unsigned char *c)
+{
+	struct uart_ra_sci_b_data *data = dev->data;
+	R_SCI_B0_Type *reg = data->sci.p_reg;
```
This value is fixed, so it's a good idea to include it in the config.

The way we do it with gpio is better.


https://github.com/zephyrproject-rtos/zephyr/pull/72150/files#diff-4fd9e45da58f81a8a9761aefd7ed42633287f99d1ad17a6b230bf0fdfa73681dR33

ydamigos commented on soc/renesas/ra/Kconfig:
```diff
@@ -1,4 +1,5 @@
 # Copyright (c) 2023 TOKITA Hiroshi <tokita.hiroshi@fujitsu.com>
+# Copyright (c) 2024 Renesas Electronics Corporation
```
Why do we add a copyright without changing anything to the file?

thaoluonguw replied:
We removed it. Thank you for your feedback.

soburi commented:
There are still a few items remaining but I recognize most are being resolved.
Finally, we need to squash the commits and clean up the history.
You can do that after the issues are resolved, but please do not forget.

soburi commented on soc/renesas/ra/ra8m1/soc.h:
```diff
@@ -0,0 +1,16 @@
+ */
+
+/**
+ * @file SoC configuration macros for the Renesas RA8 family MCU
```
nitpick: `RA8` -> `RA8M1`

thaoluonguw replied:
We update it. Thank you for your feedback.

duynguyenxa committed:
tests: Exclude EK_RA8M1 from testing CPP98

The Renesas RA HAL isn't compatible with the C++98 standard so
skipping the test ek_ra8m1 on CI

Signed-off-by: Duy Phuong Hoang. Nguyen <duy.nguyen.xa@renesas.com>

soburi commented:
My final point is about the clock in the device tree.
https://github.com/zephyrproject-rtos/zephyr/pull/72150#discussion_r1623145959
Other pointing outs have been corrected well.

The point is that the device tree has a custom to "don't write what doesn't exist,"  and also you should follow that rule.
And for xtal and subclk, you need to reflect that appropriately in the fsp definitions in hal.

https://github.com/zephyrproject-rtos/hal_renesas/pull/17#discussion_r1619541265

If necessary, you can create YAML for xtal and subclk.
The important thing is to adhere to the etiquette of the device tree notation, and to align it with the code, hide the complexity with the macro part. (In other words, it's acceptable for macros to be bit complex, IMO.)

andrzej-kaczmarek commented:
@thaoluonguw HAL changes are merged, please update manifest with new SHA for hal_renesas module

thaoluonguw commented:
Helllo @carlescufi, could you please help us confirm for change request in manifest? Thank you so much.

KhiemNguyenT commented:
@cfriedt , @nashif, we would like to have TSC review and get merged to 3.7.0. Please help to add label "TSC" for this PR.
